doc/debugger.rst

```diff
@@ -1773,6 +1811,161 @@ target endianness and interpret the bytecode based on that.
+
+    duk_idx_t duk_cb_debug_request(void* udata, duk_context* ctx, duk_idx_t nvalues) {
+        /* nop */
+        return 0;
+    }
+
+The above callback simply responds with ``REP EOM`` (an empty reply) to all
+requests.  This is admittedly not very useful and a real implementation should
+process the values it receives on the stack, push its own values to send in
+reply, and return a non-negative integer indicating how many values it pushed.
+
+Here is a slightly more useful implementation::
+
+    duk_idx_t duk_cb_debug_request(void* udata, duk_context* ctx, duk_idx_t nvalues) {
+        /* Must use negative stack indices to access dvalues */
+        const char* cmdname = duk_get_string(ctx, -nvalues + 0);
+        
+        if (strcmp(cmdname, "VersionInfo") == 0) {
+            /* Return a positive integer to send a REP containing values pushed
+             * to the stack. The return value indicates how many dvalues you
+             * are including in the response.
+             */
+            
+            duk_push_string(ctx, "My Awesome Program");
+            duk_push_int(ctx, 81200);  /* ver. 8.12.0 */
+            
+            return 2;  /* 2 dvalues */
+        } else {
+            /* Return -1 to send an ERR reply. The value on top of the stack
+             * will be string coerced and used for the error text.
+             */
+            
+            duk_push_string(ctx, "unrecognized AppRequest command name");
+            return -1;
+        }
+    }
+
+If no request callback is provided at attach, AppRequest will be treated as an
+unsupported command, eliciting an ERR reply from Duktape saying so.  Such a
+target is still free to send AppNotify messages, however.
+
+As a precaution, the target should try to avoid sending structured values such
+as JS objects in notify messages as their heap pointers may become stale by the
+time the client receives them.  This is especially true for notifications sent
+while the target is running.  It's better to stick to primitives which have
+unique dvalue representations, i.e. numbers, strings and buffers.  If a
+structured value does need to be sent, it can simply be, e.g. JSON/JX encoded
+and sent as a string instead.
+
+Important notes on the request callback
+---------------------------------------
+
+The request callback is provided with a ``duk_context`` pointer with which it
+can access the value stack and is assumed to be trusted.  Specifically, there
+are certain things it **MUST NOT** do.  Specifically:
+
+* It MUST NOT attempt to access or pop any values from the top of the stack
+  beyond the ``nvalues`` it is given.
+
+* It MUST NOT assume any specific value for ``duk_get_top()`` and similar
+  primitives.  In practice this means using negative stack indices to access
+  values.
+
+* It MUST NOT throw errors.  It is very easy to accidentally throw an error
+  when working with value stack values directly, so caution must be exercised
+  here.
```

Comment:
Maybe add:

* ... must be exercised here. (Allowing callback errors will be added in future work.)

or something similar. It's an obvious issue to fix.